Add sriov chart

manuelbuil opened the pull request:
Signed-off-by: Manuel Buil <mbuil@suse.com>

jiaqiluo commented on packages/rancher-sriov/charts/templates/_helpers.tpl:
```diff
@@ -0,0 +1,7 @@
+{{- define "system_default_registry" -}}
+{{- if .Values.global.systemDefaultRegistry -}}
```
```suggestion
{{- if .Values.global.cattle.systemDefaultRegistry -}}
```

manuelbuil replied:
good catch!

jiaqiluo commented on packages/rancher-sriov/charts/templates/view-role.yaml:
```diff
@@ -0,0 +1,11 @@
+  - sriovnetwork.openshift.io
+  resources:
+  - '*'
+  verbs:
+  - ["get", "watch", "list"]
```
```suggestion
  verbs:
  - "get"
  - "watch"
  - "list"

```

manuelbuil replied:
done

aiyengar2 commented on packages/rancher-sriov/charts/Chart.yaml:
```diff
@@ -0,0 +1,21 @@
+  - name: Rancher Labs
+    email: charts@rancher.com
+annotations:
+  catalog.cattle.io/certified: rancher
```
Just to confirm, does this need a custom UI? e.g. `catalog.cattle.io/ui-component`

manuelbuil replied:
I don't think so

aiyengar2 commented on packages/rancher-sriov/charts/Chart.yaml:
```diff
@@ -0,0 +1,21 @@
+    email: charts@rancher.com
+annotations:
+  catalog.cattle.io/certified: rancher
+  catalog.cattle.io/experimental: true
```
Does this need `catalog.cattle.io/display-name`?

manuelbuil replied:
According to confluence `  catalog.cattle.io/display-name: "Feature Name" # Only for Charts with custom UI`, as sriov does not have a custom UI, I guess we don't need that

aiyengar2 commented on packages/rancher-sriov/charts/app-README.md:
```diff
@@ -0,0 +1,13 @@
+This chart is based on the upstream [k8snetworkplumbingwg/sriov-network-operator](https://github.com/k8snetworkplumbingwg/sriov-network-operator) project. The chart deploys the SR-IOV Operator and its CRDs. It allows users to deploy SR-IOV CNI, which can be used, together with Multus, to provide high performing extra network interfaces to pods.
```
Needs title? Multus should also have a link IMO.

```suggestion
# Rancher SR-IOV Network Operator

This chart is based on the upstream [k8snetworkplumbingwg/sriov-network-operator](https://github.com/k8snetworkplumbingwg/sriov-network-operator) project. The chart deploys the SR-IOV Operator and its CRDs. It allows users to deploy SR-IOV CNI, which can be used with Multus to provide high performing extra network interfaces to pods.
```

aiyengar2 commented on packages/rancher-sriov/charts/app-README.md:
```diff
@@ -0,0 +1,13 @@
+    SR-IOV Operator - The operator helps provision and configure SR-IOV CNI plugin and Device plugin. It requires Multus to operate correctly.
+    SR-IOV Network Config Daemon - Deployed by the operator, this daemon is responsible for discovering the SR-IOV NICs on each node
+
+Note that SR-IOV requires NICs that support SR-IOV and the activation of specific configuration options in the operating system. Nodes that fultill these requirements should be labeled with:
```
nit:
```suggestion
Note that SR-IOV requires NICs that support SR-IOV and the activation of specific configuration options in the operating system. Nodes that fulfill these requirements should be labeled with:
```

aiyengar2 commented on packages/rancher-sriov/charts/crds/sriov-network-operator-sriovibnetworks_crd.yaml:
```diff
@@ -0,0 +1,74 @@
+apiVersion: apiextensions.k8s.io/v1
```
Is it possible to add a `# Source: ` to the upstream?

aiyengar2 commented on packages/rancher-sriov/package.yaml:
```diff
@@ -0,0 +1,9 @@
+url: local
+packageVersion: 01
+releaseCandidateVersion: 00
```
No longer necessary
```suggestion
```

aiyengar2 commented on packages/rancher-sriov/package.yaml:
```diff
@@ -0,0 +1,9 @@
+url: local
```
Why are we doing this as a local chart? This chart exists in an upstream https://github.com/k8snetworkplumbingwg/sriov-network-operator/tree/9b59e4ed4785bddeaea67837bd8506163a37a713/deployment/sriov-network-operator.

IMO, we should provide the Github Repository based URL for this chart and maintain just the necessary patches on top of it, unless there is a strong reason why we should not do so.

manuelbuil replied:
That's a great point. I think the person that started this work missed that there was already a helm chart upstream (possibly the work started before the upstream guys released it). I have just modified my commit and now it is based on the upstream helm chart (following the confluence docs). I tested it and works but please review ;)

aiyengar2 reviewed (changes requested):
Did not do a deep review, would like to get clarification on why we aren't declaring this as a Git-based forked chart first. https://github.com/k8snetworkplumbingwg/sriov-network-operator/tree/9b59e4ed4785bddeaea67837bd8506163a37a713/deployment/sriov-network-operator

This would be highly preferable as we only need to maintain a patch on upstream.

jiaqiluo commented on packages/rancher-sriov/generated-changes/patch/templates/clusterrolebinding.yaml.patch:
```diff
@@ -0,0 +1,18 @@
+--- charts-original/templates/clusterrolebinding.yaml
++++ charts/templates/clusterrolebinding.yaml
```
Is the patch name for the `charts/templates/clusterrolebinding.yaml` file?  
It more looks like for the `charts/templates/clusterrole.yaml` file according to the contents. 
Can you double check with it?

manuelbuil replied:
Good catch :). It's a bug upstream. I noticed it while creating this patch and fixed it upstream ==> https://github.com/k8snetworkplumbingwg/sriov-network-operator/pull/138, but it is not yet in the release we are using

jiaqiluo replied:
oh, interesting bug 😄 

We need to fix in our patch file, otherwise the `make prepare` will fail 
```
➜  charts git:(sriov-chart2) ✗ PACKAGE=rancher-sriov make prepare
./bin/charts-build-scripts prepare
INFO[0000] Pulling k8snetworkplumbingwg/sriov-network-operator@v1.0.0[path=deployment/sriov-network-operator] from upstream into charts 
INFO[0020] Loading dependencies for chart               
INFO[0020] Applying changes from generated-changes      
INFO[0020] Applying: generated-changes/patch/Chart.yaml.patch 
INFO[0020] Applying: generated-changes/patch/templates/NOTES.txt.patch 
INFO[0020] Applying: generated-changes/patch/templates/_helpers.tpl.patch 
INFO[0020] Applying: generated-changes/patch/templates/clusterrolebinding.yaml.patch 
ERRO[0020] 
patching file templates/clusterrolebinding.yaml
Hunk #1 FAILED at 15.
1 out of 1 hunk FAILED -- saving rejects to file templates/clusterrolebinding.yaml.rej 
FATA[0020] Encountered error while preparing main chart: Encountered error while trying to apply changes to charts: Unable to generate patch with error: exit status 1 
make: *** [prepare] Error 1
```

jiaqiluo commented on packages/rancher-sriov/generated-changes/exclude/crds/k8s.cni.cncf.io_networkattachmentdefinitions_crd.yaml:
```diff
@@ -0,0 +1,57 @@
+# Copyright 2020 NVIDIA
```
I see `network-attachment-definition` is listed in the clusterrole `sriov-network-operator`. 
It is ok to include it in the clusterrole even the CRD does not exist. I just need to confirm that we do want to exclude this CRD.

jiaqiluo commented on packages/rancher-sriov/generated-changes/exclude/crds/k8s.cni.cncf.io_networkattachmentdefinitions_crd.yaml:
```diff
@@ -0,0 +1,57 @@
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  name: network-attachment-definitions.k8s.cni.cncf.io
```
I see `network-attachment-definition` is listed in the clusterrole `sriov-network-operator`. 
It is ok to include it in the clusterrole even the CRD does not exist. I just need to confirm that we do want to exclude this CRD.

manuelbuil replied:
SR-IOV requires Multus. Multus allows to set up multiple interfaces in a pod. Multus becomes the main CNI plugin and can propagate calls to different "real" CNI plugins that are registered. The way to register those CNI plugins is through the crd: network-attachment-definitions.k8s.cni.cncf.io. However, that CRD is part of Multus design and thus it should be installed by Multus, not by SR-IOV. In fact, we added a few lines to detect that and if the crd is not installed, then we ask the user to install Multus first: https://github.com/rancher/charts/pull/1211/files#diff-c0260e86c59f0bb92c1629b68aad1d2f97629b9d4d1978783d810ce3470acb02R1. If we install that CRD via SR-IOV, then SR-IOV will start but we will not be able to properly allow pods to use it because we will be missing Multus binary and config which is installed by Multus helm chart. BTW, this is documented

jiaqiluo commented:
@manuelbuil Can you rebuild the charts by using the `charts-build-scripts` version v0.2.0? 

You can update it by running `make pull-scripts` locally.
```
> make pull-scripts
./scripts/pull-scripts
Pulling in charts-build-scripts version https://github.com/rancher/charts-build-scripts.git@v0.2.0
charts-build-scripts version v0.2.0 (cae97e4)
```

The rebuild the charts by running `PACKAGE=rancher-sriov make charts`.
The chart version in the final chart will be `0.1.1+up0.1.0`.

aiyengar2 commented on packages/rancher-sriov/generated-changes/patch/values.yaml.patch:
```diff
@@ -0,0 +1,52 @@
+   nameOverride: ""
+   fullnameOverride: ""
+-  resourcePrefix: "openshift.io"
++  resourcePrefix: "rancher.io"
```
What are the implications of changing the resourcePrefix? Can we safely do this without running into any other issues (e.g. CRDs still have openshift.io on them right?)

manuelbuil replied:
Yes, this is safe. CRDs still have the openshift.io suffix. The sriov resources are named $resourcePrefix/$nic_model, e.g. rancher.io/intel

cbron commented on charts/rancher-sriov/sriov/1.0.1/Chart.yaml:
```diff
@@ -0,0 +1,21 @@
+name: sriov
+sources:
+- https://github.com/rancher/charts
+version: 1.0.1
```
Needs icon in here pointing at https://charts.rancher.io/assets/logos/.....

manuelbuil commented on packages/rancher-sriov/generated-changes/overlay/templates/admin-role.yaml:
```diff
@@ -0,0 +1,12 @@
+  name: sriov-admin
+rules:
+- apiGroups:
+  - sriovnetwork.openshift.io
```
As explained above, I'd not do it because it is not part of SR-IOV but of Multus

aiyengar2 replied:
So going back to the explanation in https://github.com/rancher/charts/pull/1211#discussion_r639711405, even if we are not installing that CRD as part of this chart, should a user who has sriov-admin permissions be able to also register and unregister CNI plugins or do we expect that to be a permission that Multus will need to provide?

manuelbuil commented on packages/rancher-sriov/generated-changes/overlay/templates/admin-role.yaml:
```diff
@@ -0,0 +1,12 @@
+apiVersion: rbac.authorization.k8s.io/v1
```
Yes, I can do that

aiyengar2 replied:
Yeah I agree with putting it in a clusterrole.yaml, but if would be nice if we could put this under `templates/rancher-sriov` similar to what we do with Monitoring here https://github.com/rancher/charts/tree/dev-v2.5/charts/rancher-monitoring/rancher-monitoring/14.5.101%2Bup14.5.0/templates/rancher-monitoring.

It's way nicer to maintain an overlay file that a big patch 😅 , but clusterrole.yaml makes more sense to contain clusterroles that role.yaml.

aiyengar2 replied:
Oh I see you already added it. Can you drop these overlay files then?

manuelbuil commented on packages/rancher-sriov/generated-changes/patch/templates/operator.yaml.patch:
```diff
@@ -0,0 +1,70 @@
+       containers:
+         - name: {{ include "sriov-network-operator.fullname" . }}
+-          image: {{ .Values.images.operator }}
++          image: {{ include "system_default_registry" . }}{{ .Values.images.operator }}
```
It's ugly but it's in the value already:
```
images:
  operator: rancher/hardened-sriov-network-operator:v1.0.0-build20210429
 ```

aiyengar2 replied:
I think we should patch this chart to have the values.yaml be more consistent with other charts, w.r.t. the image tag, such as:

```
images:
  operator:
    image: rancher/hardened-sriov-network-operator
    tag: v1.0.0-build20210429
    os: linux # optional, but if we eventually support windows this will be necessary for airgap too
```

We can then patch all references here accordingly, e.g. 
```suggestion
+          image: {{ include "system_default_registry" . }}{{ .Values.images.operator.image }}{{ .Values.images.operator.tag }}
```

manuelbuil commented on index.yaml:
```diff
@@ -3282,6 +3282,48 @@ entries:
+    - https://github.com/rancher/charts
+    urls:
+    - assets/rancher-sriov/sriov-1.0.1.tgz
+    version: 1.0.1
```
I recreated index. yaml but it seems there are two packages, right?
```
    urls:    - assets/rancher-sriov/sriov-1.0.1.tgz    version: 1.0.1
```
and
```
    urls:    - assets/rancher-sriov/sriov-0.1.1+up0.1.0.tgz    version: 0.1.1+up0.1.0
```
should I remove the one without `up`?

jiaqiluo replied:
Yes, you need to remove the one without `up` in the following places:
- the index.yaml file
- the `assets/rancher-riov` directory

jiaqiluo reviewed (changes requested):
overall LGTM, we just need to do the following changes:
- remove assets/rancher-sriov/sriov-1.0.1.tgz
- remove assets/rancher-sriov/sriov-crd-1.0.1.tgz
- update the `index.yaml`

aiyengar2 commented on packages/rancher-sriov/generated-changes/overlay/templates/secrets.yaml:
```diff
@@ -0,0 +1,20 @@
+metadata:
+  name: operator-webhook-service
+  namespace: {{ .Release.Namespace }}
+data: {{ include "sriov_operator_cert" . | indent 2 }}
```
need the newline right?
```suggestion
data: {{ include "sriov_operator_cert" . | nindent 2 }}
```

aiyengar2 commented on packages/rancher-sriov/generated-changes/overlay/templates/secrets.yaml:
```diff
@@ -0,0 +1,20 @@
+metadata:
+  name: network-resources-injector-secret
+  namespace: {{ .Release.Namespace }}
+data: {{ include "sriov_resource_injector_cert" . | indent 2 }}
```
```suggestion
data: {{ include "sriov_resource_injector_cert" . | nindent 2 }}
```

manuelbuil commented:
> overall LGTM, we just need to do the following changes:
> 
>     * remove assets/rancher-sriov/sriov-1.0.1.tgz
> 
>     * remove assets/rancher-sriov/sriov-crd-1.0.1.tgz
> 
>     * update the `index.yaml`

Done! Thanks a lot for your patience and comprehensive reviews

aiyengar2 commented on charts/rancher-sriov/sriov/0.1.1+up0.1.0/.helmignore:
```diff
@@ -0,0 +1,23 @@
+# Patterns to ignore when building packages.
```
Assets need to be regenerated. `0.1.1` is incorrect, right?

manuelbuil replied:
The upstream version is 0.1.0 --> https://github.com/k8snetworkplumbingwg/sriov-network-operator/blob/master/deployment/sriov-network-operator/Chart.yaml#L3. Not sure how is the last 1 filled, any idea?

aiyengar2 replied:
Gotcha. So the fix is to the CRD chart version, not to this chart. Will leave a comment for that in my next review, but resolving this.

aiyengar2 commented on packages/rancher-sriov/generated-changes/overlay/app-README.md:
```diff
@@ -0,0 +1,14 @@
+
+The chart installs the following components:
+
+ - SR-IOV Operator - The operator helps provision and configure SR-IOV CNI plugin and Device plugin. It requires Multus to operate correctly.
+ - SR-IOV Network Config Daemon - Deployed by the operator, this daemon is responsible for discovering the SR-IOV NICs on each node
```
nit: even if you don't take these changes, make sure you either add punctuation to the second bullet point or remove it from the first.

```suggestion
 - SR-IOV Operator - An operator that helps provision and configure CNI or Device plugins (requires Multus)
 - SR-IOV Network Config Daemon - A Daemon deployed by the Operator that discovers SR-IOV NICs on each node
```

manuelbuil replied:
I partially applied your suggestion. I think we should have `provision and configure SR-IOV CNI plugin and SR-IOV Device plugin`. If we remove SR-IOV, the user could think that any CNI or any Device Plugin can be deployed.

What do you mean by adding punctuation to the second bullet?

aiyengar2 replied:
You resolved the issue. I mean that the first had a `.` at the end and the second did not 😄

aiyengar2 commented on packages/rancher-sriov/generated-changes/overlay/app-README.md:
```diff
@@ -0,0 +1,14 @@
+# Rancher SR-IOV Network Operator
+
+This chart is based on the upstream [k8snetworkplumbingwg/sriov-network-operator](https://github.com/k8snetworkplumbingwg/sriov-network-operator) project. The chart deploys the SR-IOV Operator and its CRDs. It allows users to deploy SR-IOV CNI, which can be used, together with [Multus](https://github.com/k8snetworkplumbingwg/multus-cni), to provide high performing extra network interfaces to pods ...
```
```suggestion
This chart is based on the upstream [k8snetworkplumbingwg/sriov-network-operator](https://github.com/k8snetworkplumbingwg/sriov-network-operator) project. The chart deploys the SR-IOV Operator and its CRDs, which is designed to help the user to provision and configure SR-IOV CNI and Device plugins in a cluster that uses [Multus CNI](https://github.com/k8snetworkplumbingwg/multus-cni) to provide high performing extra network interfaces to pods.
```

aiyengar2 commented on packages/rancher-sriov/generated-changes/overlay/templates/certmanagercerts.yaml:
```diff
@@ -0,0 +1,43 @@
+{{- if .Values.operator.enableAdmissionController -}}
+{{- if .Values.cert_manager -}}
+{{- if not (.Capabilities.APIVersions.Has "cert-manager.io/v1") -}}
```
Are these intentionally separated or could we use an `and` here to omit having three `end` blocks?

manuelbuil replied:
good idea. I can remove one end combining the first two

aiyengar2 commented on packages/rancher-sriov/generated-changes/overlay/templates/certmanagercerts.yaml:
```diff
@@ -0,0 +1,43 @@
+apiVersion: cert-manager.io/v1
+kind: Issuer
+metadata:
+  name: sriov-network-operator-selfsigned-issuer
```
Should the naming scheme for these be based on `{{ include "sriov-network-operator.fullname" . }}`?

manuelbuil replied:
I did not want to add another delta with regards to upstream. I'd suggest we wait until upstream implements this

manuelbuil replied:
You are right, this is currently us supporting it. There are a couple of PRs upstream adding the certmanager stuff and they are not yet merged. Until then, we must support it ourselves. I have just looked at why it does the `{{ include "sriov-network-operator.fullname" . }}` instead of adding the string directly. I see that it is just in case there is a fullnameOverride in the values. We could add this here too but I don't see a big benefit of doing it

aiyengar2 replied:
If we are going to deprecate our certmanager stuff in favor of an eventual upstream change, I think it's fine to leave it as-is. Resolving this comment.

aiyengar2 commented on packages/rancher-sriov/generated-changes/overlay/templates/secrets.yaml:
```diff
@@ -0,0 +1,20 @@
+apiVersion: v1
+kind: Secret
+metadata:
+  name: operator-webhook-service
```
Should the naming scheme for these be based on `{{ include "sriov-network-operator.fullname" . }}`?

manuelbuil replied:
I did not want to add another delta with regards to upstream. I'd suggest we wait until upstream implements this

manuelbuil replied:
Same as above. If you believe this is important, I could spend time adding new functions in _helpers but I don't see the benefit

aiyengar2 commented on packages/rancher-sriov/generated-changes/patch/templates/clusterrole.yaml.patch:
```diff
@@ -0,0 +1,61 @@
++apiVersion: rbac.authorization.k8s.io/v1
++kind: ClusterRole
++metadata:
++  name: sriov-admin
```
Do we expect these clusterroles to automatically aggregate into the default k8s admin, edit, and view roles? If so, we need annotations here, e.g.

https://github.com/rancher/charts/blob/7836aff0c1b2846c3fbcf8ed2adc6963d21c20b1/charts/rancher-monitoring/rancher-monitoring/14.5.101%2Bup14.5.0/templates/rancher-monitoring/clusterrole.yaml#L7-L9

manuelbuil replied:
Not sure about this. I added this roles because it was required based on https://confluence.suse.com/pages/viewpage.action?spaceKey=EN&title=Feature+Charts+Helm+Requirements, i.e. we don't need them to run this chart correctly. Do you normally recommend to add that to charts?

aiyengar2 replied:
So the reason why these annotations are necessary is because they [aggregate these ClusterRoles](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#aggregated-clusterroles) to the [default k8s roles](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#user-facing-roles), which are bound to corresponding Rancher role ([Project Owner -> admin](https://github.com/rancher/rancher/blob/9c9491b2879de82454ab53d083492e530e519693/pkg/data/management/role_data.go#L229), [Project Member -> edit](https://github.com/rancher/rancher/blob/9c9491b2879de82454ab53d083492e530e519693/pkg/data/management/role_data.go#L265), [Read-only -> view](https://github.com/rancher/rancher/blob/9c9491b2879de82454ab53d083492e530e519693/pkg/data/management/role_data.go#L297)).

This is a choice made for every chart, so I think it depends on your RBAC expectations for this chart. Do you expect the person who owns or is part of the project that contains `cattle-sriov-system` to have these permissions by default, or should they be manually bound as and when necessary?

cc: @cbron if you have any thoughts.

manuelbuil replied:
Thanks for the explanation. I can see users having only "sriov-view" permissions, because this chart modifies the network configuration of the host and that can create connectivity issues across nodes. Therefore, only the "admin" of the cluster should be able to have full control of this. I guess these should be made available by default. Let me add those

aiyengar2 reviewed (changes requested):
Still need some more changes but the finish line is in sight! 😄  

Also, IMO it might make sense to temporarily split your changes into two commits (one for `packages` and the other for the output of `make charts`) and squash once we have finished reviews. That way you can just redo the last commit after we have reviewed the `packages` changes and squash it into the first.

aiyengar2 commented on packages/rancher-sriov/templates/crd-template/Chart.yaml:
```diff
@@ -0,0 +1,11 @@
+apiVersion: v2
+version: 1.0.0
```
Version should match v0.1.0 or I think the `catalog.cattle.io/auto-install: sriov-crd=match` annotation will break

https://github.com/k8snetworkplumbingwg/sriov-network-operator/blob/07398ed56de01f6e7d2a869743c6d5ce8298ca5b/deployment/sriov-network-operator/Chart.yaml#L3

aiyengar2 reviewed (changes requested):
Need to regenerate assets once you get the CRD chart versioning change in.

aiyengar2 commented on packages/rancher-sriov/generated-changes/patch/templates/clusterrole.yaml.patch:
```diff
@@ -0,0 +1,54 @@
++kind: ClusterRole
++metadata:
++  name: sriov-edit
++  {{- if .Values.global.rbac.userRoles.aggregateToDefaultRoles }}
```
Does a default need to be patched in the values.yaml? e.g.

```yaml
global:
  rbac:
    userRoles:
      aggregateToDefaultRoles: true|false
```

If a `helm template` does not cause a failure and the default for this is expected to be false, then we can continue to omit this. But I'm not sure what would happen since `global.rbac.userRoles` is undefined right now.

manuelbuil replied:
Yes, we need this

aiyengar2 replied:
Sounds good.

aiyengar2 commented on index.yaml:
```diff
@@ -3386,6 +3386,67 @@ entries:
+    - assets/rancher-sriov/sriov-0.1.1+up0.1.0.tgz
+    version: 0.1.1+up0.1.0
+  sriov-crd:
+  - annotations:
+      catalog.cattle.io/certified: rancher
+      catalog.cattle.io/experimental: "true"
+      catalog.cattle.io/hidden: "true"
+      catalog.cattle.io/namespace: cattle-sriov-system
+      catalog.cattle.io/release-name: sriov-crd
+    apiVersion: v2
+    created: "2021-05-27T21:06:06.912831309+02:00"
+    description: Installs the CRDs for rke2-sriov.
+    digest: ef9e1ad98796590a04447a094a399e41cba27bc48bf5172d58c1f95a675e5a03
+    name: sriov-crd
+    type: application
+    urls:
+    - assets/rancher-sriov/sriov-crd-1.0.1+up0.1.0.tgz
+    version: 1.0.1+up0.1.0
```
should not exist?

aiyengar2 commented:
@manuelbuil Thank you for being patient with all the requested changes 😄 

I'm **almost** ready to approve this PR. Before I do, can you:
- [ ] Rebase to the latest upstream (we just fixed CI workflows for charts)
- [ ] Remove the `index.yaml`, `charts/`, and `asset/` changes entirely from the commit and re-run `make charts` to avoid issues like https://github.com/rancher/charts/pull/1211#discussion_r643386272.
- [ ] Verify `helm template assets/rancher-sriov/sriov-crd-0.1.1+up0.1.0.tgz` is successful and prints out a k8s manifest containing just your CRDs
- [ ] Verify `helm template assets/rancher-sriov/sriov-0.1.1+up0.1.0.tgz` is successful and prints out a k8s manifest corresponding to the expected resources (to ensure something like https://github.com/rancher/charts/pull/1211#discussion_r643384352 does not cause an issue)
- [ ] (Optional, but recommended) Verify that a helm install of the chart onto a Rancher downstream cluster with Multus is successful. You can do this by navigating to the Apps & Marketplace UI of the downstream cluster, adding a Charts Repository on Rancher with url `https://github.com/manuelbuil/charts` and branch `sriov-chart2`, and trying to install the chart.

Let's get this merged! 🎉

manuelbuil commented:
> @manuelbuil Thank you for being patient with all the requested changes smile
> 
> I'm **almost** ready to approve this PR. Before I do, can you:
> 
>     * [x]  Rebase to the latest upstream (we just fixed CI workflows for charts)
> 
>     * [x]  Remove the `index.yaml`, `charts/`, and `asset/` changes entirely from the commit and re-run `make charts` to avoid issues like [#1211 (comment)](https://github.com/rancher/charts/pull/1211#discussion_r643386272).
> 
>     * [x]  Verify `helm template assets/rancher-sriov/sriov-crd-0.1.1+up0.1.0.tgz` is successful and prints out a k8s manifest containing just your CRDs
> 
>     * [x]  Verify `helm template assets/rancher-sriov/sriov-0.1.1+up0.1.0.tgz` is successful and prints out a k8s manifest corresponding to the expected resources (to ensure something like [#1211 (comment)](https://github.com/rancher/charts/pull/1211#discussion_r643384352) does not cause an issue)
> 
>     * [ ]  (Optional, but recommended) Verify that a helm install of the chart onto a Rancher downstream cluster with Multus is successful. You can do this by navigating to the Apps & Marketplace UI of the downstream cluster, adding a Charts Repository on Rancher with url `https://github.com/manuelbuil/charts` and branch `sriov-chart2`, and trying to install the chart.
> 
> 
> Let's get this merged! tada

Thank you for you patience!

I had an old deployment of rancherd v2.5.8. I added `https://github.com/manuelbuil/charts` with branch `sriov-chart2` and I see sriov-crd and sriov. When I try to install any of them I get:

```
NotFound (404)

catalogtemplateversions.management.cattle.io "myrepo-sriov-crd-0.1.1+up0.1.0" not found
```
and
```
NotFound (404)

catalogtemplateversions.management.cattle.io "myrepo-sriov-0.1.1+up0.1.0" not found
```
Does this point to an error in my PR? Or is it because using rancherd v2.5? BTW, is it normal to get both helm charts (crd and non-crd)?

jiaqiluo commented on charts/rancher-sriov/sriov/0.1.1+up0.1.0/values.yaml:
```diff
@@ -0,0 +1,54 @@
+operator:
+  tolerations:
+  - effect: NoSchedule
+    key: node-role.kubernetes.io/control-plane
```
```suggestion
    key: node-role.kubernetes.io/controlplane
```

jiaqiluo commented on charts/rancher-sriov/sriov/0.1.1+up0.1.0/values.yaml:
```diff
@@ -0,0 +1,54 @@
+    operator: Equal
+    value: linux
+  nodeSelector:
+    node-role.kubernetes.io/control-plane: "true"
```
```suggestion
    node-role.kubernetes.io/controlplane: "true"
```

manuelbuil commented:
We tried to deploy with @jiaqiluo using rancher 2.5 and realized that taint and label changes between rke1 and rke2 from `node-role.kubernetes.io/controlplane` to `node-role.kubernetes.io/control-plane`. We were wondering if there is a way to only show the app when the cluster is rke2 and how we could test it with rancher 2.6 and rke2. rancher 2.5 only deploys rke1

aiyengar2 commented:
> I had an old deployment of rancherd v2.5.8. I added `https://github.com/manuelbuil/charts` with branch `sriov-chart2` and I see sriov-crd and sriov. When I try to install any of them I get:
> 
> ```
> NotFound (404)
> 
> catalogtemplateversions.management.cattle.io "myrepo-sriov-crd-0.1.1+up0.1.0" not found
> ```
> 
> and
> 
> ```
> NotFound (404)
> 
> catalogtemplateversions.management.cattle.io "myrepo-sriov-0.1.1+up0.1.0" not found
> ```

Where are you seeing this issue on an install? I was able to trigger an install on your chart successfully and get the CRDs installed on a downstream cluster in Rancher 2.5.8, although it failed shy of installing the full chart since I didn't deploy it on a cluster with Multus installed. Does the failure you encountered happen after this step?

![Screen Shot 2021-06-02 at 12 12 29 PM](https://user-images.githubusercontent.com/13127443/120538857-cf49ce80-c39b-11eb-8430-126b5ff93bd7.png)

> Does this point to an error in my PR? Or is it because using rancherd v2.5? BTW, is it normal to get both helm charts (crd and non-crd)?

My assumption is **no**, your PR looks fine. Maybe it could be because you are using rancherd v2.5, but I'm not sure. It's possible that you needed to manually refresh the catalog before proceeding with the install for some reason.

It is normal to get both Helm charts since our UI automatically installs the CRD chart before the main chart.

aiyengar2 commented:
> We tried to deploy with @jiaqiluo using rancher 2.5 and realized that taint and label changes between rke1 and rke2 from `node-role.kubernetes.io/controlplane` to `node-role.kubernetes.io/control-plane`. We were wondering if there is a way to only show the app when the cluster is rke2 and how we could test it with rancher 2.6 and rke2. rancher 2.5 only deploys rke1

To my knowledge, there is no way to specify that a chart is only shown to specific cluster types. However, we could introduce a custom UI that overrides the above values based on the cluster type; this is something we do for Monitoring today.

aiyengar2 commented on packages/rancher-sriov/generated-changes/overlay/app-README.md:
```diff
@@ -0,0 +1,14 @@
+# Rancher SR-IOV Network Operator
+
+This chart is based on the upstream [k8snetworkplumbingwg/sriov-network-operator](https://github.com/k8snetworkplumbingwg/sriov-network-operator) project. The chart deploys the SR-IOV Operator and its CRDs, which is designed to help the user provision and configure the SR-IOV CNI in a cluster that uses [Multus CNI](https://github.com/k8snetworkplumbingwg/multus-cni), to provide high performing ex ...
```
Based on offline conversation, we need to make sure this file calls out that:
1) This chart is expected to be deployed on an RKE2 cluster. It has not been tested or designed for any other cluster type at the moment, although it can be deployed in other clusters with some modifications. (Addresses https://github.com/rancher/charts/pull/1211#issuecomment-853288767)
2) A user who is deploying this chart is expected to be fairly knowledgable about Multus, SR-IOV, and the use case behind this chart as it is only meant for **advanced use cases** where multiple CNIs and high performing network interfaces on pods are required. Users who do not need these features are not advised to install this chart.

cc: @cjellick @cbron

jiaqiluo replied:
In the case of the sriov chart, although we can fix the pod tolerations to make it installed successfully in a rke1 cluster,  I think we will keep it as it is today so the failure can also bring the user’s attention and realize the cluster is not an RKE2 cluster.

aiyengar2 reviewed (changes requested):
Need to make the changes Jack suggested for nodeSelectors, specifically for RKE 2, and re-check this list: https://github.com/rancher/charts/pull/1211#issuecomment-852359393.

Also, we need to make an app-README.md change due to offline conversation.

manuelbuil commented:
> > I had an old deployment of rancherd v2.5.8. I added `https://github.com/manuelbuil/charts` with branch `sriov-chart2` and I see sriov-crd and sriov. When I try to install any of them I get:
> > ```
> > NotFound (404)
> > 
> > catalogtemplateversions.management.cattle.io "myrepo-sriov-crd-0.1.1+up0.1.0" not found
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and
> > ```
> > NotFound (404)
> > 
> > catalogtemplateversions.management.cattle.io "myrepo-sriov-0.1.1+up0.1.0" not found
> > ```
> 
> Where are you seeing this issue on an install? I was able to trigger an install on your chart successfully and get the CRDs installed on a downstream cluster in Rancher 2.5.8, although it failed shy of installing the full chart since I didn't deploy it on a cluster with Multus installed. Does the failure you encountered happen after this step?
> 
> ![Screen Shot 2021-06-02 at 12 12 29 PM](https://user-images.githubusercontent.com/13127443/120538857-cf49ce80-c39b-11eb-8430-126b5ff93bd7.png)
> 
> > Does this point to an error in my PR? Or is it because using rancherd v2.5? BTW, is it normal to get both helm charts (crd and non-crd)?
> 
> My assumption is **no**, your PR looks fine. Maybe it could be because you are using rancherd v2.5, but I'm not sure. It's possible that you needed to manually refresh the catalog before proceeding with the install for some reason.
> 
> It is normal to get both Helm charts since our UI automatically installs the CRD chart before the main chart.

I was using rancherd v2.5, that was the issue. Once I moved to the newer verison, it worked :)

manuelbuil commented:
>     * sriov-chart2



> @manuelbuil Thank you for being patient with all the requested changes smile
> 
> I'm **almost** ready to approve this PR. Before I do, can you:
> 
>     * [x]  Rebase to the latest upstream (we just fixed CI workflows for charts)
> 
>     * [x]  Remove the `index.yaml`, `charts/`, and `asset/` changes entirely from the commit and re-run `make charts` to avoid issues like [#1211 (comment)](https://github.com/rancher/charts/pull/1211#discussion_r643386272).
> 
>     * [x]  Verify `helm template assets/rancher-sriov/sriov-crd-0.1.1+up0.1.0.tgz` is successful and prints out a k8s manifest containing just your CRDs
> 
>     * [x]  Verify `helm template assets/rancher-sriov/sriov-0.1.1+up0.1.0.tgz` is successful and prints out a k8s manifest corresponding to the expected resources (to ensure something like [#1211 (comment)](https://github.com/rancher/charts/pull/1211#discussion_r643384352) does not cause an issue)
> 
>     * [ ]  (Optional, but recommended) Verify that a helm install of the chart onto a Rancher downstream cluster with Multus is successful. You can do this by navigating to the Apps & Marketplace UI of the downstream cluster, adding a Charts Repository on Rancher with url `https://github.com/manuelbuil/charts` and branch `sriov-chart2`, and trying to install the chart.
> 
> 
> Let's get this merged! tada
![image](https://user-images.githubusercontent.com/16151814/120672206-62653000-c492-11eb-9162-0dbdcebf3019.png)

jiaqiluo commented on packages/rancher-sriov/package.yaml:
```diff
@@ -0,0 +1,10 @@
+url: https://github.com/k8snetworkplumbingwg/sriov-network-operator.git
+subdirectory: deployment/sriov-network-operator
+commit: c5c943cc214f4113b6f6be8d7c48ac837aa1669c
+packageVersion: 01
```
nit. 

```suggestion
packageVersion: 1
```

jiaqiluo reviewed (dismissed):
LGTM

aiyengar2 commented on packages/rancher-sriov/generated-changes/overlay/app-README.md:
```diff
@@ -0,0 +1,13 @@
+# Rancher SR-IOV Network Operator
+
+This chart is based on the upstream [k8snetworkplumbingwg/sriov-network-operator](https://github.com/k8snetworkplumbingwg/sriov-network-operator) project. The chart deploys the SR-IOV Operator and its CRDs, which is designed to help the user provision and configure the SR-IOV CNI in a cluster that uses [Multus CNI](https://github.com/k8snetworkplumbingwg/multus-cni), to provide high performing ex ...
```
nit:
```suggestion
This chart is based on the upstream [k8snetworkplumbingwg/sriov-network-operator](https://github.com/k8snetworkplumbingwg/sriov-network-operator) project. The chart deploys the SR-IOV Operator and its CRDs, which are designed to help the user provision and configure the SR-IOV CNI in a cluster that uses [Multus CNI](https://github.com/k8snetworkplumbingwg/multus-cni), to provide high performing extra network interfaces to pods. This chart is expected to be deployed on an RKE2 cluster and only meant for advanced use cases where multiple CNI plugins and high performing network interfaces on pods are required. Users who do not need these features are not advised to install this chart.
```

manuelbuil replied:
Thanks

aiyengar2 commented on packages/rancher-sriov/generated-changes/overlay/app-README.md:
```diff
@@ -0,0 +1,13 @@
+
+The chart installs the following components:
+
+ - SR-IOV Operator - An operator that helps provision and configure the SR-IOV CNI plugin and SR-IOV Device plugin (requires Multus)
```
nit: `requires Multus` is also mentioned above. Do we need this?
```suggestion
 - SR-IOV Operator - An operator that helps provision and configure the SR-IOV CNI plugin and SR-IOV Device plugin
```

aiyengar2 commented on packages/rancher-sriov/generated-changes/patch/templates/role.yaml.patch:
```diff
@@ -0,0 +1,7 @@
+       - configmaps
+     verbs:
+       - get
++
```
nit: do we need this file if it is just adding a newline?

aiyengar2 reviewed (approved):
LGTM. I have a couple of nitpicks below that should be addressed before merging (specifically to the app-README.md), but I'm good to sign off on this. Can approve again if you make the changes.

Thanks again for making all the changes! Let's get this merged 🎉

aiyengar2 commented:
Also, @manuelbuil, unless I overlooked the relevant PSPs, we are assuming that this initial version of the chart will **not** be deployable on a hardened cluster that pass a CIS 1.16 scan right? 

I don't think we should introduce any more fixes into this PR, but we may need additional changes at some point to support that use case. If a ticket does not currently exist, we should open one up.

manuelbuil commented:
> Also, @manuelbuil, unless I overlooked the relevant PSPs, we are assuming that this initial version of the chart will **not** be deployable on a hardened cluster that pass a CIS 1.16 scan right?
> 
> I don't think we should introduce any more fixes into this PR, but we may need additional changes at some point to support that use case. If a ticket does not currently exist, we should open one up.

Yes, you are right. I haven't take into account anything related to CIS. When you say ticket, do you mean an issue in this repo?

cbron reviewed (commented):
Few minor things, basically gtg besides permission aggregation situation

manuelbuil commented on charts/rancher-sriov/sriov/0.1.1+up0.1.0/Chart.yaml:
```diff
@@ -0,0 +1,25 @@
+description: SR-IOV network operator configures and manages SR-IOV networks in the
+  kubernetes cluster
+home: https://github.com/k8snetworkplumbingwg/sriov-network-operator
+icon: https://charts.rancher.io/assets/logos/sr-iov.svg
```
You mean in `remotes/origin/release-v2.6`? What should I put there, the result of doing `PACKAGE=rancher-sriov make charts`, i.e. what gets created in charts/ and assets/?

cbron replied:
Oh sorry I just meant the logo svg file, so it doesn't 404, ex: https://charts.rancher.io/assets/logos/logging.svg

manuelbuil replied:
I cherry-picked it into release-v2.6 branch: https://github.com/rancher/charts/pull/1269

manuelbuil commented on charts/rancher-sriov/sriov/0.1.1+up0.1.0/values.yaml:
```diff
@@ -0,0 +1,57 @@
+    tag: v1-build20210310
+  webhook:
+    image: rancher/hardened-sriov-network-webhook
+    tag: v1.0.0-build20210429
```
Yes, they are all hardened. It's basically taking the upstream code but building them using our golang build image that uses goboring

manuelbuil commented:
@cbron The configuring of some of the sriov crds ends up configuring hardware resources from the nodes. That is pretty intrusive, so I'd suggest only the cluster admin to be capable of doing that. In that case, should I remove the:
```
  {{- if .Values.global.rbac.userRoles.aggregateToDefaultRoles }}
    rbac.authorization.k8s.io/aggregate-to-admin: "true"
  {{- end }}
```
from sriov-adming and sriov-edit?

manuelbuil commented:
> @cbron The configuring of some of the sriov crds ends up configuring hardware resources from the nodes. That is pretty intrusive, so I'd suggest only the cluster admin to be capable of doing that. In that case, should I remove the:
> 
> ```
>   {{- if .Values.global.rbac.userRoles.aggregateToDefaultRoles }}
>     rbac.authorization.k8s.io/aggregate-to-admin: "true"
>   {{- end }}
> ```
> 
> from sriov-adming and sriov-edit?

Clarified with Caleb. Admin, edit and view will have the same permissions because we want only the cluster admin to set the config of sr-iov
